### PR TITLE
Ability System

### PR DESCRIPTION
**Pull request by Tdead1**
Made a small tag-based ability system inspired by how UE4 manages it, please take a look.

### CODE REVIEW

File: source/Subsystem/EntityComponentSystem.h [hunk 1]
```diff
@@ -2,27 +2,28 @@
...
 	MODEL = 2
 };
 
-typedef unsigned int UniqueID;
-
 class EntityComponentSystem
 {
 public:
-	EntityComponentSystem() : myUIDCounter(0) {};
+	EntityComponentSystem() : myUIDCounter(0), myAbilitySystem(this) {};
```
[the remark this answers is not in the document]
Comment (Tdead1) [reply source/Subsystem/EntityComponentSystem.h:22]: Nope, just being constructed after a forward declare :)

File: source/Subsystem/AbilitySystem.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#include <functional>
+
+#include "../Utility/TypeDefines.h"
+
+class EntityComponentSystem;
+
+enum class AbilityTag
+{
+	Any = 0
+	, Player
```
[the remark this answers is not in the document]
Comment (Tdead1) [reply source/Subsystem/AbilitySystem.h:12]: It's easier to expand on, but sure, I'll change it to default 👍

File: source/Subsystem/AbilitySystem.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
+#include "AbilitySystem.h"
+#include "EntityComponentSystem.h"
+
+bool AbilitySystem::StartAbilityOnEntity(Ability anAbility, const UniqueID aTargetEntity)
+{
+	Entity* affectedEntity = myEntityComponentSystem->GetEntity(aTargetEntity);
```
Comment (TDCRanila) [inline source/Subsystem/AbilitySystem.cpp:6]: affectedEntity can be const.

File: source/Subsystem/AbilitySystem.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	if (!affectedEntity || !anAbility.myEffect)
+		return false;
+
+	bool affectTagsMatch = false;
+	
+	for (auto tag : anAbility.myAffectTags)
```
Comment (TDCRanila) [inline source/Subsystem/AbilitySystem.cpp:12]: If I remember, you wanted to refrain from using auto for most things, so consider using the proper type here.

File: source/Subsystem/AbilitySystem.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	if (!affectTagsMatch)
+		return false;
+	
+	if (anAbility.myAffectOverridesIgnore)
+	{
+		anAbility.myEffect();
+		return true;
+	}
+
+	for (auto tag : anAbility.myIgnoreTags)
```
Comment (TDCRanila) [inline source/Subsystem/AbilitySystem.cpp:30]: If I remember, you wanted to refrain from using auto for most things, so consider using the proper type here.

File: source/Subsystem/AbilitySystem.cpp (hunk 4 above)
Comment (TDCRanila) [reply source/Subsystem/AbilitySystem.cpp:12]: Can also be const.

File: source/Subsystem/AbilitySystem.cpp (hunk 5 above)
Comment (TDCRanila) [reply source/Subsystem/AbilitySystem.cpp:30]: Can also be const.

File: source/Subsystem/AbilitySystem.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	std::vector<AbilityTag>* abilityTagList = myTagsTable[anEntityID];
+	if (!abilityTagList)
+	{
+		myTagsTable.emplace(anEntityID, nullptr);
+		myTagsTable[0] = new std::vector<AbilityTag>{ aTag };
+		return true;
+	}
+
+	bool alreadyTagged = false;
+	for (auto& i : *abilityTagList)
```
Comment (TDCRanila) [inline source/Subsystem/AbilitySystem.cpp:53]: If I remember, you wanted to refrain from using auto for most things, so consider using the proper type here. 
Can also be const.

Comment (TDCRanila) [reply source/Subsystem/AbilitySystem.cpp:53]: Before you were using "tag" for the for loops, but now you are using "i" instead. Should be consistent.

File: source/Subsystem/AbilitySystem.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+bool AbilitySystem::HasTag(const AbilityTag aTag, const UniqueID anEntityID)
+{
+	std::vector<AbilityTag>* abilityTagList = myTagsTable[anEntityID];
+	if (!abilityTagList)
+		return false;
+
+	bool alreadyAffected = false;
+	for (auto& i : *abilityTagList)
```
Comment (TDCRanila) [inline source/Subsystem/AbilitySystem.cpp:76]: If I remember, you wanted to refrain from using auto for most things, so consider using the proper type here.
Can also be const.
Before you were using "tag" for the for loops, but now you are using "i" instead. Should be consistent.

File: source/Subsystem/AbilitySystem.cpp (hunk 3 above)
Comment (Tdead1) [reply source/Subsystem/AbilitySystem.cpp:6]: Fixed.

File: source/Subsystem/AbilitySystem.cpp (hunk 4 above)
Comment (Tdead1) [reply source/Subsystem/AbilitySystem.cpp:12]: Fixed.

File: source/Subsystem/AbilitySystem.cpp (hunk 5 above)
Comment (Tdead1) [reply source/Subsystem/AbilitySystem.cpp:30]: Fixed.

File: source/Subsystem/AbilitySystem.cpp (hunk 6 above)
Comment (Tdead1) [reply source/Subsystem/AbilitySystem.cpp:53]: Fixed.

File: source/Subsystem/AbilitySystem.cpp (hunk 7 above)
Comment (Tdead1) [reply source/Subsystem/AbilitySystem.cpp:76]: Fixed.

**Pull request merged**